arangod/Transaction/Methods.cpp

```diff
@@ -409,1386 +409,1607 @@ OperationResult emptyResult(OperationOptions const& options) {
+        }
+        // otherwise build a proper error message
+        Result result{TRI_ERROR_ARANGO_UNIQUE_CONSTRAINT_VIOLATED};
+        auto index = _collection.getPhysical()->primaryIndex();
+        if (index) {
+          return index->addErrorMsg(result, key);
+        }
+        return result;
+      } else {
+        TRI_ASSERT(lookupResult.first.isSet());
+        TRI_ASSERT(lookupResult.second.isSet());
+        std::tie(oldDocumentId, oldRevisionId) = lookupResult;
+      }
+    } else if (res.errorNumber() != TRI_ERROR_ARANGO_DOCUMENT_NOT_FOUND) {
+      // Error reporting in the babies case is done outside of here.
+      if (res.is(TRI_ERROR_ARANGO_CONFLICT) && !isArray) {
+        TRI_ASSERT(_replicationType != Methods::ReplicationType::FOLLOWER);
+        // if possible we want to provide information about the conflicting
+        // document, so we need another lookup. However, it is possible that
+        // the other transaction has not been committed yet, or that the
+        // document has been deleted by now.
+        auto lookupRes = _collection.getPhysical()->lookupKey(
+            &_methods, key, lookupResult, ReadOwnWrites::yes);
+        TRI_ASSERT(!lookupRes.ok() || lookupResult.second.isSet());
+        _methods.buildDocumentIdentity(_collection, _resultBuilder,
+                                       _trxColl.id(), key, lookupResult.second,
+                                       RevisionId::none(), nullptr, nullptr);
+        auto index = _collection.getPhysical()->primaryIndex();
+        if (index) {
+          return index->addErrorMsg(res, key);
+        }
+        return res;
+      }
+      return res;
     }
 
-    collection = trxCollection(cid);
-    if (collection == nullptr) {
-      throwCollectionNotFound(collectionName);
-    }
+    std::ignore = _methods.state()->ensureSnapshot();
 
-  } else {
-    AccessMode::Type collectionAccessType = collection->accessType();
-    if (AccessMode::isRead(collectionAccessType) && !AccessMode::isRead(type)) {
-      THROW_ARANGO_EXCEPTION_MESSAGE(
-          TRI_ERROR_TRANSACTION_UNREGISTERED_COLLECTION,
-          std::string(
-              TRI_errno_string(TRI_ERROR_TRANSACTION_UNREGISTERED_COLLECTION)) +
-              ": " + collectionName + " [" + AccessMode::typeString(type) +
-              "]");
-    }
-  }
+    // only populated for update/replace
+    transaction::BuilderLeaser previousDocumentBuilder(&_methods);
 
-  TRI_ASSERT(collection != nullptr);
-  return cid;
-}
+    RevisionId newRevisionId;
+    bool didReplace = false;
+    bool excludeFromReplication = _excludeAllFromReplication;
 
-/// @brief add a collection to the transaction for read, at runtime
-DataSourceId transaction::Methods::addCollectionAtRuntime(
-    std::string const& collectionName, AccessMode::Type type) {
-  if (collectionName == _collectionCache.name && !collectionName.empty()) {
-    return _collectionCache.cid;
-  }
+    if (not oldDocumentId.isSet()) {
```

Comment:
Our current convention (used in >= 99.9999% of our code) is to use `!` and not `not` (pun intended).
```suggestion
    if (!oldDocumentId.isSet()) {
```

Reply:
I think I saw somewhere in the code `not` instead if `!` which is the only reason I used it. :)
I actually prefer `not` because I find it much more readable - a `!` can be overlooked very easily.